src/viper/trans.ml

```diff
@@ -66,19 +86,37 @@ let rec unit (u : M.comp_unit) : prog =
         it = [], init_list;
         note = NoInfo
       } in
-    let m = 
-      { it = MethodI(init_id, [
-          self_id, { it = RefT; at = self_id.at; note = NoInfo }
-        ], [], [], [], Some init_body);
+    let init_m =
+      { it = MethodI(init_id, [self_id, self_typ], [], [], [], Some init_body);
         at = no_region;
         note = NoInfo
       } in
-    let is = m :: is' in
-    let invs = extract_invariants is in
-      { it = adorn_invariants invs is;
-        at = body.at;
-        note = NoInfo
-      }
+    let is'' = init_m :: is' in
+    (* Add permissions *)
+    let is''' = List.map (fun {it; at; note: info} -> (
+      match it with
+      | MethodI (id, ins, outs, pres, posts, body) ->
+        { at;
+          it = MethodI (id, ins, outs, (perm at)::pres, (perm at)::posts, body);
+          note
+        }
+        | _ -> {it; at; note}
+    )) is'' in
+    (* Add functional invariants *)
+    let invs = extract_invariants is''' (self_id, self_typ) [] in
+    let is = List.map (fun {it; at; note: info} ->
```

Comment:
maybe you can use `List.mapi` to observe the first element, and treat is specially w.r.t. the precondition.